File: 5-longest-palindromic-substring/longest-palindromic-substring.py
def expand(s,i,j):
    if s[i]!=s[j]: return ""
    while i>=0 and j<len(s) and s[i] == s[j]:
        i-=1 
        j+=1
    return s[i+1:j]

class Solution:
    def longestPalindrome(self, s: str) -> str:
        n=len(s)
        res=''
        for i in range(n):
            res = max(res, expand(s,i,i), key=len)
            if i<n-1: res = max(res, expand(s,i,i+1), key=len)
        return res

File: 5-longest-palindromic-substring/longest-palindromic-substring.cpp
string preprocess(string s){
    string res = "#";
    for(char c : s){
        res += c;
        res += '#';
    }
    return res;
}

vector<int> manachers(string s){
    s = preprocess(s);
    int n = s.size();
    vector<int> res(n);
    int l = 0, r = 0, mir = 0; 
    for (int idx = 0; idx<s.size(); idx++){
        if (idx >= r || idx + res[2*mir - idx] >= r){
            mir = idx;
            r = max(r, idx);
            l = 2*idx - r;
            while (0<=l-1 && r+1<n && s[l-1] == s[r+1]){
                l--;
                r++;
            }
            res[idx] = r-idx;
        } else {
            int x = 2*mir - idx;
            res[idx] = res[x];
        }
    }

    return res;
}

class Solution {
public:
    string longestPalindrome(string s) {
        vector<int> res = manachers(s);
        int idx = max_element(res.begin(), res.end())-res.begin();
        int maxlen = idx + res[idx];
        int start = idx - res[idx];
        s = preprocess(s);
        string result = "";
        for (int i = start+1; i < maxlen; i+=2){
            result += s[i];
        }
        return result;
    }
};